### PR TITLE
Limit rate of requests in rest server

### PR DESCRIPTION
**Pull request by kashizui**
Limits unauthenticated users (identified by IP) to 60 requests per hour, and authenticated users to 5,000 requests per hour.

Uses the same HTTP header conventions as [the GitHub API](https://developer.github.com/v3/#rate-limiting).

### CODE REVIEW

File: codalab/server/rest_server.py [hunk 1]
```diff
@@ -41,6 +43,77 @@
...
+
+            # Update count and prevent integer overflow
+            self._counts[user_id] = (min(count, limit), first_time)
+
+            # Returning remaining count and (truncated) remaining time
+            return limit - count, time_remaining
+
+
+class RateLimitPlugin(object):
+    """
```
Comment (klopyrev) [inline codalab/server/rest_server.py:79]: Make a note that this is per-process. We usually run with multiple processes.

File: codalab/server/rest_server.py [hunk 2]
```diff
@@ -41,6 +43,77 @@
 ]
 
 
+class RequestCounter(object):
```
Comment (klopyrev) [inline codalab/server/rest_server.py:46]: I have no idea why you're doing this. If someone really wanted to DDOS us they'd get around your REMOTE_ADDR check.

Do you expect some existing code to send a ton of requests?

File: codalab/server/rest_server.py [hunk 3]
```diff
@@ -41,6 +43,77 @@
...
+                user_id = request.environ.get('REMOTE_ADDR')
+                limit = self.unauth_limit
+
+            remaining, reset = self.request_counter.count(user_id, limit)
+            response.set_header('X-RateLimit-Limit', limit)
+            response.set_header('X-RateLimit-Remaining', remaining)
+            response.set_header('X-RateLimit-Reset', reset)
+            if remaining < 0:
+                # Cannot use abort, since it ignores the headers on response
+                response.status = FORBIDDEN
```
Comment (klopyrev) [inline codalab/server/rest_server.py:110]: Code 429 seems more appropriate.

File: codalab/server/rest_server.py [hunk 4]
```diff
@@ -41,6 +43,77 @@
...
+    """
+    Limit the request rate.
+    Request counts are associated with user for authenticated requests,
+    and with IP address for unauthenticated requests. Different limits can be
+    configured for these two cases.
+    Sets the appropriate headers following the GitHub API's conventions.
+    """
+    api = 2
+
+    def __init__(self, window_seconds=60*60, auth_limit=5000, unauth_limit=60):
```
Comment (klopyrev) [inline codalab/server/rest_server.py:88]: 60 requests per hour for unauthorized seems too low. What if someone is browsing through the website, looking at the different bundles. I can see them hit 60 requests in a few minutes. With 4 processes, that's 240. Still, a few minutes of  heavy browsing would hit that.

Comment (klopyrev) [inline codalab/server/rest_server.py:88]: A worker can easily hit the 5000 per hour limit. I don't support this change at all. I'd just not do this change and worry about too many requests when we start seeing too many requests.

Comment (kashizui) [inline codalab/server/rest_server.py:88]: Ah that's fair. That and the multiprocess thing and the weak REMOTEADDR check I think are the nails in the coffin for this one.

**Comment by kashizui** [comment]
Pigeonholing until the need arises, in which case this will require a much more robust implementation...

**Pull request closed without merging**